docs/design/2024-01-22-ticdc-large-txn-not-block-wm.md

```diff
@@ -0,0 +1,205 @@
+
+## Solution: large transactions periodically update its min-commit-ts
+
+From "How large transactions block the advancing of Resolve TS" section we can see the root cause of large transactions block resolve ts is:
+
+- Current Resolve TS(water mark) calculation depends on the start ts of all ongoing transactions, and the start ts of a large transaction is relevant "old" and not advanced when the large transaction is running.
+  More reasonable choice is using min-commit-ts of large transactions to calculate the water mark, and same as the TTL, the large transaction periodically update its primary key's min-commit-tsinformation to let the water mark mechanism can advance the water mark smoothly. The following diagram demostrates the basic idea of this new proposal:
+
+![image](../media/large-txn-no-block-wm-3.png)
+
+## Detailed Design
+
+### Large Transaction Write
+
+There are few code changes for the large transaction write part. As mentioned above, everytime the ongoing large transaction is going to update its primary key's TTL information, the transaction fetches the latest TSO from PD, and then updates the TS as min-commit-ts to the primary key.
+
+- All locks add a "large transaction" flag.
+- Large transaction periodically update primary key's min-commit-ts when updating TTL, to indicate that this large transaction is still going on and its commit ts must greater than this min-commit-ts.
+
+```
+WatermarkAdvancer (former named Resolver in TiKV)
+struct WatermarkAdvancer {
+    // A region represents a data range
+    region_id: u64,
+
+    // key -> start_ts, only record for normal transactions
+    locks_by_key: HashMap<Arc<[u8]>, TimeStamp>,
+    // start_ts -> locked keys, only record for normal transactions
+    lock_start_ts_heap: BTreeMap<TimeStamp, HashSet<Arc<[u8]>>>,
+
+    // Ongoing large transactions related info
+    // start_ts -> (min-commit-ts, primary key) for ongoing large transactions
+    large_txn_map: HashMap<TimeStamp, (TimeStamp, primary key)>
+    // (min-commit-ts, start_ts) set, in order to distinguish different large txns with
+    // the same min-commit-ts value
+    large_txn_min_commit_ts_heap: BTreeSet<(TimeStamp,TimeStamp)>
+}
+```
+
+We separate normal transactions and large transactions in the TS Resolver module. Because each lock contains the transaction-size (or we expliciltely add large-transaction flag in the lock) information, when the Lock Observer receives a lock operation, we can dispatch the lock to related region's Resolver and according to the large transaction flag. We just store a little information like start-ts, primary-key, update-to-date min-commit-ts in the Resolver for a large transaction even though there are 1M rows changed in a region. This can significantly reduce the total memory usage of Resolver for a large transaction which can reduce the OOM issue of TiKV (we previously met some OOM issue because the Resolver occupies a lot of memory for large transactions).
```

Comment:
```suggestion
We separate normal transactions and large transactions in the TS Resolver module. Because each lock contains the transaction-size (or we explicitly add large-transaction flag in the lock) information, when the Lock Observer receives a lock operation, we can dispatch the lock to related region's Resolver and according to the large transaction flag. We just store a little information like start-ts, primary-key, update-to-date min-commit-ts in the Resolver for a large transaction even though there are 1M rows changed in a region. This can significantly reduce the total memory usage of Resolver for a large transaction which can reduce the OOM issue of TiKV (we previously met some OOM issue because the Resolver occupies a lot of memory for large transactions).
```